internal/yunikorn/sync.go

```diff
@@ -94,55 +95,117 @@ func (s *Service) upsertPartitions(ctx context.Context) ([]*dao.PartitionInfo, e
 	return partitions, nil
 }
 
-// upsertPartitionQueues fetches queues for each partition and upserts them into the database
-func (s *Service) upsertPartitionQueues(ctx context.Context, partitions []*dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
+// syncQueues fetches queues for each partition and upserts them into the database
+func (s *Service) syncQueues(ctx context.Context, partitions []*dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
 	logger := log.FromContext(ctx)
 
-	// Create a wait group as a separate goroutine will be spawned for each partition
-	wg := sync.WaitGroup{}
-	wg.Add(len(partitions))
+	var errs []error
+	var allQueues []*dao.PartitionQueueDAOInfo
+	var mu sync.Mutex
+	var wg sync.WaitGroup
 
-	// Protect the queues and errors slices with a mutex so multiple goroutines can safely append queues
-	mutex := sync.Mutex{}
+	addErr := func(err error) {
+		mu.Lock()
+		defer mu.Unlock()
+		errs = append(errs, err)
+	}
 
-	var errs []error
-	var queues []*dao.PartitionQueueDAOInfo
+	addQueues := func(queues []*dao.PartitionQueueDAOInfo) {
+		mu.Lock()
+		defer mu.Unlock()
+		allQueues = append(allQueues, queues...)
+	}
+
+	addJob := func(partition *dao.PartitionInfo) {
+		wg.Add(1)
+		jobName := fmt.Sprintf("sync_queues_for_partition_%s", partition.Name)
+
+		err := s.workqueue.Add(func(ctx context.Context) error {
+			defer wg.Done()
+
+			queues, err := s.syncPartitionQueues(ctx, partition)
+			if err != nil {
+				addErr(fmt.Errorf("error syncing partition queues for partition %s: %w", partition.Name, err))
+			} else {
+				addQueues(queues)
+			}
+			return nil
+		}, workqueue.WithJobName(jobName))
 
-	processPartition := func(p *dao.PartitionInfo) {
-		defer wg.Done()
-		queue, err := s.client.GetPartitionQueues(ctx, p.Name)
-		mutex.Lock()
-		defer mutex.Unlock()
 		if err != nil {
-			errs = append(errs, fmt.Errorf("could not get queues for partition %s: %v", p.Name, err))
-		} else {
-			queues = append(queues, queue)
+			logger.Errorf("could not add sync queues job for partition %s: %v", partition.Name, err)
 		}
 	}
 
-	for _, p := range partitions {
-		go processPartition(p)
+	for _, partition := range partitions {
+		addJob(partition)
```

Comment:
I think we are complicating too much here. I'm not sure what do we gain from pushing sync of queues to the background. The sync partitions is pushed to the background. That is a single unit of work. Then, correctly synchronized partition is the one that synchronized the queue. Since we are waiting on the workqueue, it is effectively the same as waiting on the goroutine. So why not use the language primitive?

The idea of the workqueue is (from my point of view) to do something in the background. Usually, best-effort cleanups and such should go there. But since we care about the result, we should simply leverage goroutines to make our life easier. The way we used workqueue here is the same way we would use the goroutine:

```go
errs := make(chan error, len(partitions))
partitionQueues := make(chan []*dao.PartitionQueueDAOInfo, len(partitions))
wg.Add(len(partitions))
for _, partition := range partitions {
  go func() {
    defer wg.Done()
    queues, err := syncPartitionQueues(partition)
    if err != nil {
      errs <- err
      return
    }
    partitionQueues <- queues
  }()
}
wg.Wait()
close(errs)
close(partitionQueues)
var syncErrors []error
for err := range errs {
  syncErrors = append(syncErrors, err)
}
if len(syncErrors) > 0 {
  return errors.Join(syncErrors...)
}
var queues []*dao.PartitionQueueDAOInfo
for qs := range partitionQueues {
  queues = append(queues, qs...)
}
return queues, nil
```